### PR TITLE
structMap[@TYPE=OCR-D-LOGICAL] / FULLDOWNLOAD

### PR DESCRIPTION
**Pull request by kba**

### CODE REVIEW

File: mets.md [hunk 1]
```diff
@@ -160,6 +199,83 @@ encodings of the same page.
 </mets:structMap>
 ```
 
+## OCR-D structMap
+
+A METS can contain different `structMap`. These are defined by the `TYPE` attribute (e.g. `LOGICAL`, `PHYSICAL`>) described in more detail.
```
Comment (bertsky) [inline mets.md:204]:
```suggestion
A METS may contain different `mets:structMap` entries, differentiated by their `TYPE` attribute (e.g. `LOGICAL`, `PHYSICAL`). Some of these are mandatory within OCR-D.
```

File: mets.md [hunk 2]
```diff
@@ -160,6 +199,83 @@ encodings of the same page.
 </mets:structMap>
 ```
 
+## OCR-D structMap
+
+A METS can contain different `structMap`. These are defined by the `TYPE` attribute (e.g. `LOGICAL`, `PHYSICAL`>) described in more detail.	
+A document structure determined by OCR-D is described by `TYPE="OCR-D-LOGICAL"`.
```
Comment (bertsky) [inline mets.md:205]: What exactly does this mean? There's also the mandatory `PHYSICAL` layout. So maybe this should read: 
```suggestion
The _logical_ document structure detected by MP MUST be described by `TYPE="OCR-D-LOGICAL"`.
```

File: mets.md [hunk 3]
```diff
@@ -160,6 +199,83 @@ encodings of the same page.
...
+
+attributes in `structMap` | description
+--                        | --
+`LABEL`                   | contains the character recognition result  e.g. the title of the chapter
```
Comment (bertsky) [inline mets.md:209]:
```suggestion
`LABEL`                   | contains the recognized text of a structuring element, e.g. the title of the chapter
```

File: mets.md [hunk 4]
```diff
@@ -160,6 +199,83 @@ encodings of the same page.
...
+`TYPE`                    | contains a standardized description of structur (see.: [DFG-Viewer: structural data set](https://dfg-viewer.de/strukturdatenset/))
```
Comment (bertsky) [inline mets.md:210]:
```suggestion
`TYPE`                    | contains the type of a structuring element according to a standardized, controlled vocabulary (see.: [DFG-Viewer: structural data set](https://dfg-viewer.de/strukturdatenset/))
```

File: mets.md [hunk 5]
```diff
@@ -160,6 +199,83 @@ encodings of the same page.
...
+```xml
+<mets:fileGrp USE="OCR-D-IMG">
+    <mets:file ID="OCR-D-IMG_0001" >...</mets:file>
+</mets:fileGrp>
+<mets:fileGrp USE="OCR-D-OCR">
+    <mets:file ID="OCR-D-OCR_0001" >...</mets:file>
+</mets:fileGrp>
+<mets:structMap TYPE="OCR-D-LOGICAL">
+  <mets:div DMDID="dmdSec_0001" ADMID="amdSec_0001" ID="OCR-D-loc_0001">
+    <mets:div ID="OCR-D-loc_d5e320" TYPE="chapter" LABEL="Kapidel 1">
```
Comment (bertsky) [inline mets.md:228]: Is this a typo or an intentional "recognition error"?

Comment (tboenig) [reply mets.md:228]: This is a simulated "recognition error".

Comment (bertsky) [reply mets.md:228]: I see. Then maybe this would be more realistic / clear:
```suggestion
    <mets:div ID="OCR-D-loc_d5e320" TYPE="chapter" LABEL="KapıteI 1">
```

File: mets.md (hunk 4 above)
Comment (tboenig) [reply mets.md:210]: I think the the proposal is not better.
element -> is ambiguous, and it is not an element it is an index term or subject term or descriptor
For this reason I propose:
`contains a descriptor according to a standardized, controlled vocabulary`

File: mets.md (hunk 3 above)
Comment (tboenig) [reply mets.md:209]: ->structuring element : a title of the chapter is not an element, that´s only text.
better: `contains the recognized text e.g. the title of the chapter`

File: mets.md (hunk 4 above)
Comment (bertsky) [reply mets.md:210]: My formulation doesn't just use _element_ but the compound noun _structuring element_. Both rows refer to this in my suggestion (i.e. its _text_ and its _type_). A heading or caption is of course a structuring element (within a structured document), not an `index term or subject term or descriptor`.

Comment (tboenig) [reply mets.md:211]: No, the value of `@TYPE` is an **descriptor**.

File: mets.md [hunk 6]
```diff
@@ -131,6 +137,39 @@ PROCESSOR := [A-Z0-9\-]{3,}
 `<mets:file ID="OCR-D-IMG_0001">`            | The unmanipulated source image
 `<mets:file ID="OCR-D-IMG-BIN_0001">`        | Black-and-White image
 
+#### Fulldownload
+
+The `ID` attribute prefix `FULLDOWNLOAD` followed by the file format (TEI, ALTO, hOCR, HTML, TXT, PDF) must be used for a requested file, their are the hole publication or digital book.
```
[the remark this answers is not in the document]
Comment (tboenig) [reply mets.md:143]: No, this type of formulation requires that the documentation be read sequentially first.

What is important?
I think the documentation of the attribute value, with its specifics (prefix_format) + a characteristic description (whole book ...).

My proposal:

`For `mets:file` entries representative of the publication **as a whole** (not just a single page), the `ID` attribute MUST have  prefix `FULLDOWNLOAD_`, followed by the file format (`TEI`, `ALTO`, `hOCR`, `HTML`, `TXT`, `COCO`, `PDF`).`

File: mets.md (hunk 4 above)
Comment (bertsky) [reply mets.md:211]: > No, the value of `@TYPE` is an **descriptor**.

I get a feeling this is just a misunderstanding:

Of course `@TYPE` is a descriptor (as is `@LABEL`). But they **belong** to (or refer to) a _structuring element_ like `chapter`. You'll have to explain the general thing first, however obvious it may seem. In contrast, that `@TYPE` is a descriptor is not worth extra mention IMO.

See what I mean?

File: mets.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply mets.md:143]: > No, this type of formulation requires that the documentation be read sequentially first.

Huh?

If you wanted to say that mentioning the structMap here already anticipates the respective section: of course it does, that's why it's also a hyperlink to that section! But you cannot make this fully linear if you also want to give a full example below. (And in your draft it is the next section which is really misordered unnecessarily.)

File: mets.md (hunk 1 above)
Comment (tboenig) [reply mets.md:205]: A METS-file may contain different `mets:structMap` entries, differentiated by their `TYPE` attribute (e.g. `LOGICAL`, `PHYSICAL, ...`). 

- A `mets:structMap` with `TYPE="PHYSICAL"` is mandatory.  

- The *logical* document structure detected by *library or archive* MUST be described by `TYPE="LOGICAL"`.

- The *logical* document structure detected by *OCR-D software* MUST be described by `TYPE="OCR-D-LOGICAL"`.

File: mets.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (tboenig) [reply mets.md:143]: I think you misunderstood something here.

>>No, this type of formulation requires that the documentation be read sequentially first. *referred to* Instead of the above `OCR-D-` prefix recommendation,`  That is also the reason why I would like this passage deleted.

[the remark this answers is not in the document]
Comment (bertsky) [reply mets.md:143]: > I think you misunderstood something here.

It's possible. But you are making it hard for me to understand…

(Von mir aus auch gern auf Deutsch.)

> No, this type of formulation requires that the documentation be read sequentially first. _referred to_ Instead of the above `OCR-D-` prefix recommendation,` That is also the reason why I would like this passage deleted.

Can you please use Markdown properly so I can clearly see what you quote, what you add, what you emphasize and what you refer to verbatim? It's really hard to parse this kind of mixture and you already pointed out there must have been some confusion.

File: mets.md (hunk 4 above)
Comment (tboenig) [reply mets.md:211]: I see it only now. Here is missing a description to  'mets:div'.

Comment (bertsky) [reply mets.md:211]: > Here is missing a description to 'mets:div'.

Yes, you _could_ point that out in the table header – for example by replacing
> attributes in `structMap`

with
> attributes of `mets:structMap/mets:div/mets:div`

But this is _not_ what I meant...

File: mets.md (hunk 1 above)
Comment (tboenig) [reply mets.md:205]:
```suggestion
A METS may contain different `mets:structMap` entries, differentiated by their `TYPE` attribute (e.g. `LOGICAL`, `PHYSICAL, ...`). 
* A `mets:structMap` with `TYPE="PHYSICAL"` is mandatory.
* The _logical_ document structure detected by _library or archive_ MUST be described by `TYPE="LOGICAL"`.
* The _logical_ document structure detected by _OCR-D software_ MUST be described by `TYPE="OCR-D-LOGICAL"`.
```

File: mets.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply mets.md:143]: Ok, so after direct consultation we identified the source of confusion. @tboenig is concerned that formulations like 
> Instead of the above `OCR-D-` prefix recommendation

(directed at text coherence) will require a linear reading flow, which is not appropriate for a specification. Instead, both subsections (the previous one about file ID naming for single pages and the new one on `FULLDOWNLOAD`) should be on the same level, and self-contained.

[the remark this answers is not in the document]
Comment (bertsky) [reply mets.md:143]:
```suggestion
For `mets:file` entries representative of the publication **as a whole**, the `ID` attribute MUST have  prefix `FULLDOWNLOAD_`, followed by the file format (`TEI`, `ALTO`, `hOCR`, `HTML`, `TXT`, `COCO`, `PDF`).

These entries SHOULD be referenced in the [structMap](#ocr-d-structmap) under `/mets:mets/mets:structMap[@TYPE="PHYSICAL"]/mets:fptr`.
```

[the remark this answers is not in the document]
Comment (bertsky) [reply mets.md:143]: I cannot add suggestions for the previous subsection (because it is outside of the diff context), but roughly I would:
- introduce a level-3 heading beneath `File ID syntax` for either single-page files (with `OCR-D-...` convention) or whole-document files (with `FULLDOWNLOAD...` convention)
- add a note in the former that physical pages must be referenced in the physical structMap (also with a hyperlink)

File: mets.md (hunk 1 above)
Comment (bertsky) [reply mets.md:205]: Maybe you could add a hyperlink to the METS profile for logical structMap we want to use here?

File: mets.md (hunk 4 above)
Comment (bertsky) [reply mets.md:211]: > > attributes of `mets:structMap/mets:div/mets:div`
> 
> But this is _not_ what I meant...

So we agreed that the recursive nature  of these `mets:div` should be mentioned as well.

But the culprit here was that @tboenig is concerned the term _element_ gets confused with its technical reading in the context of the XML standard.

But since we somehow still have to say just what a structMap consists of (i.e. what `@TYPE` and `@LABEL` describe), I proposed to use the term _(structuring) component_ instead.

This would become:
```suggestion
`TYPE` | contains the type of a structuring component according to some standardized, controlled vocabulary (see.: [DFG-Viewer: structural data set](https://dfg-viewer.de/strukturdatenset/))
```

File: mets.md (hunk 3 above)
Comment (bertsky) [reply mets.md:210]: See discussion below.

```suggestion
`LABEL` | contains the recognized text of a structuring component, e.g. the title of a chapter
```

File: mets.md [hunk 7]
```diff
@@ -95,12 +95,19 @@ with the type of manipulation (`BIN-KRAKEN`).
...
 `<mets:fileGrp USE="OCR-D-SEG-WORD">`       | Word segmentation
 `<mets:fileGrp USE="OCR-D-SEG-GLYPH">`      | Glyph segmentation
 `<mets:fileGrp USE="OCR-D-OCR-TESS">`       | Tesseract OCR
 `<mets:fileGrp USE="OCR-D-OCR-ANY">`        | AnyOCR
+`<mets:fileGrp USE="OCR-D-TEI">`            | TEI
+`<mets:fileGrp USE="OCR-D-ALTO">`           | ALTO
+`<mets:fileGrp USE="OCR-D-hOCR">`           | hOCR
+`<mets:fileGrp USE="OCR-D-HTML">`           | HTML
+`<mets:fileGrp USE="OCR-D-TXT">`            | Text
+`<mets:fileGrp USE="OCR-D-COCO">`            | COCO JSON
```
Comment (bertsky) [inline mets.md:109]:
```suggestion
`<mets:fileGrp USE="OCR-D-COCO">`           | [COCO](http://cocodataset.org/#format-data)
```

File: mets.md (hunk 2 above)
Comment (tboenig) [reply mets.md:211]:
```suggestion
The _logical_ document structure detected by OCR-D software MUST be described by `TYPE="OCR-D-LOGICAL"`.
```

Comment (bertsky) [reply mets.md]: I'd be okay with either, but AFAICS the term _MP_ (for module project) has been used more often so far.

Maybe we should revisit this, anyway – the spec should just call them "implementors" or "implementing software" or "OCR-D software"

File: mets.md [hunk 8]
```diff
@@ -160,6 +199,83 @@ encodings of the same page.
 </mets:structMap>
 ```
 
+## OCR-D structMap
```
[the remark this answers is not in the document]
Comment (tboenig) [reply mets.md:207]: here my proposal for a new document structure:

**Requirements on handling METS/PAGE**
1. Metadata
1.1 Unique `@ID` for the document processed

2. Images
2.1. Pixel density of images must be explicit and high enough
2.2. No multi-page images
2.3  Image coordinates
2.4  If in PAGE then in METS


3. File Group `mets:fileGrp `
3.1 `@USE` syntax
Examples
4. File `mets:file`
4.1` @ID` syntax
Examples
4.2 `@MIMETYPE` syntax
Examples
Examples (Media Type for PAGE XML)



5. Grouping files by page `mets:structMap`
Example
5.1 `@TYPE` syntax
Example

6. Range of pages `mets:structLink`
Example

7. Paths
7.1 Always use URL or relative filenames
Example


8. Recording processing information in METS

[the remark this answers is not in the document]
Comment (bertsky) [reply mets.md:207]: In that case, I would:
- subsume 8 (processing information) under 1 (metadata)
- abandon 2.3 (frankly, I don't know why this resides here and not just in PAGE.md)
- replace 2.3 with a general note about original/derived images (what is now in [PAGE.md](https://ocr-d.de/en/spec/page#url-for-imagefilename--filename), but including new language from #164)

But I wonder: where in that outline did `Fulldownload` go? Is it still subsumed under 4.1 for you? (We discussed this elsewhere: then you cannot make these subsections self-contained.)

[the remark this answers is not in the document]
Comment (tboenig) [reply mets.md:207]: > * abandon 2.3 (frankly, I don't know why this resides here and not just in PAGE.md)
> * replace 2.3 with a general note about original/derived images (what is now in [PAGE.md](https://ocr-d.de/en/spec/page#url-for-imagefilename--filename), but including new language from #164)

That's right, I think 2.3 is better found in page.md.

> * subsume 8 (processing information) under 1 (metadata)
 That`s a good proposal. 

> But I wonder: where in that outline did `Fulldownload` go? Is it still subsumed under 4.1 for you? (We discussed this elsewhere: then you cannot make these subsections self-contained.)

> But I wonder: where in that outline did `Fulldownload` go? Is it still subsumed under 4.1 for you? (We discussed this elsewhere: then you cannot make these subsections self-contained.)

Yes, `Fulldownload` is a section/part under 4.1

[the remark this answers is not in the document]
Comment (tboenig) [reply mets.md:207]: Requirements on handling METS/PAGE

1 Metadata
1.1 Recording processing information in METS
1.2 Unique `@ID` for the document processed

2 Images
2.1. Pixel density of images must be explicit and high enough
2.2. No multi-page images 
2.3 If in PAGE then in METS

3 File Group `mets:fileGrp`
3.1 `@USE` syntax
Examples
3.2 `@USE="FULLDOWNLOAD_..."`
Examples

4 File `mets:file`
4.1 `@ID` syntax
Examples
4.2 `@MIMETYPE` syntax
Examples
Examples (Media Type for PAGE XML)

5 Grouping files by page `mets:structMap`
Example
5.1 `@TYPE` syntax
Example

6. Range of pages `mets:structLink`
Example

Paths
7.1 Always use URL or relative filenames
Example

File: mets.md [hunk 9]
```diff
@@ -160,6 +204,86 @@ encodings of the same page.
...
+
+A METS may contain different `mets:structMap` entries, differentiated by their `TYPE` attribute (e.g. `LOGICAL`, `PHYSICAL, ...`). 
+* A `mets:structMap` with `TYPE="PHYSICAL"` is mandatory.
+* The _logical_ document structure detected by _library or archive_ MUST be described by `TYPE="LOGICAL"`.
+* The _logical_ document structure detected by _OCR-D software_ MUST be described by `TYPE="OCR-D-LOGICAL"`.
+The _logical_ document structure detected by OCR-D software MUST be described by `TYPE="OCR-D-LOGICAL"`.
```
Comment (bertsky) [inline mets.md:213]:
```suggestion
```

File: mets.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kba) [reply mets.md:192]: Is this resolved?

**Review by cneud — CHANGES_REQUESTED** [review]
I would agree with @bertsky to implement these changes:

> replace 2.3 with a general note about original/derived images (what is now in PAGE.md, but including new language from #164)

> The section `Fulldownload` should still be part of the file ID syntax (differentiating between page-local and document-global naming scheme, but not trying to formulate this "self-contained"). 

> Also, the section about `grouping files by structMap` should come below fileGrp and file ID sections.

I think also 7.1 could be 1.3 instead.

File: mets.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply mets.md:192]: > I think also 7.1 could be 1.3 instead.

Yes!

**Review by tboenig — APPROVED** [review]

**Comment by kba** [comment]
Superseded by https://github.com/OCR-D/spec/pull/207, @tboenig please have a look there.
